sources/web/datalab/jupyter.ts

```diff
@@ -81,67 +88,78 @@ function pipeOutput(stream: NodeJS.ReadableStream, port: number, error: boolean)
  * Starts the Jupyter server, and then creates a proxy object enabling
  * routing HTTP and WebSocket requests to Jupyter.
  */
-function createJupyterServer(userId: string): JupyterServer {
+function createJupyterServer(userId: string, resolved: (server: JupyterServer)=>void, rejected:()=>void) {
```

Comment:
Moving the code for the 'resolved' and 'rejected' methods into this one has been done, but putting the logic into a for loop has not.

Specifically, there's an issue with trying to use a for loop, as this method is using some other (externally defined) methods that are already defined to be asynchronous, so we can't use them in a blocking fashion.